# __Pipeline 3:__
## _IQR --> Normalization --> Z Score_
## _Min Max --> Normalization --> 3 STD_

## Importing Data:

In [1]:
# import required package for data handling
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib as plt
import matplotlib.pyplot as plt
from pandas import DataFrame
%matplotlib inline

# import required packages for splitting data
from sklearn import model_selection
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import train_test_split

# import required packages for evaluating models
from sklearn import metrics
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
from sklearn.metrics import roc_auc_score
from sklearn.metrics import accuracy_score
from sklearn.metrics import precision_recall_fscore_support

# import `logistic regression` model
from sklearn.linear_model import LogisticRegression

In [2]:
compdata_original = pd.read_csv('comp1data.csv', header=0, na_values = '-')
compdata_original.head()

I1                                  I2    I3  P(IPO)  P(H)  P(L)  \
0  AATI  ADVANCED ANALOGIC TECHNOLOGIES INC  3674    10.0   9.5   8.5   
1  ABPI     ACCENTIA BIOPHARMACEUTICALS INC  2834     8.0  10.0   8.0   
2  ACAD          ACADIA PHARMACEUTICALS INC  2834     7.0  14.0  12.0   
3  ACHN       ACHILLION PHARMACEUTICALS INC  2834    11.5  16.0  14.0   
4  ACLI     AMERICAN COMMERCIAL LINES INC.   4492    21.0  21.0  19.0   

     P(1Day)     C1   C2    C3  ...          C6       C7     T1       T2  \
0  11.870000  122.0  1.0  3.43  ...  10600000.0   51.345  470.0  12719.0   
1   7.250000  259.0  0.0 -1.62  ...   2400000.0   25.936  791.0  21792.0   
2   6.700000   90.0  1.0 -1.24  ...   5000000.0    7.378  201.0   5262.0   
3  12.390000  209.0  1.0 -0.91  ...   4500000.0    8.526  328.0   8259.0   
4  56.599998   80.0  1.0  0.07  ...   8250000.0  632.298  572.0  14830.0   

        T3     T4      T5    S1     S2     S3  
0  11560.0  301.0   690.0  62.0  117.0  139.0  
1  19585.0  510.0  1120.0  71.0  242.0  237.0  
2   4785.0  128.0   325.0  61.0   33.0   60.0  
3   7574.0  177.0   509.0  80.0   59.0  110.0  
4  13176.0  336.0   720.0  67.0  149.0  167.0  

[5 rows x 22 columns]

# Descriptive Statistics

In [3]:
compdata_original.shape
# 682 rows and 22 columns

(682, 22)

In [4]:
compdata_original.describe()

P(IPO)        P(H)        P(L)      P(1Day)           C1  \
count  677.000000  672.000000  672.000000   660.000000   660.000000   
mean    13.837666   15.481190   13.515045    25.934766   149.728788   
std      6.053731    6.653429    5.835646    73.234948   152.817467   
min      3.000000    0.000000    3.000000     0.000000    10.000000   
25%     10.000000   12.500000   11.000000    11.000000    85.000000   
50%     13.500000   15.000000   13.000000    14.845000   107.000000   
75%     17.000000   17.000000   15.000000    20.485000   155.250000   
max     85.000000  135.000000  108.000000  1159.200562  2087.000000   

               C2           C3          C4            C5            C6  \
count  660.000000   646.000000  660.000000  6.760000e+02  6.760000e+02   
mean     0.859091     1.788904    0.007282  4.935776e+07  1.241519e+07   
std      0.348192   162.666532    0.033318  1.043764e+08  2.512855e+07   
min      0.000000  -786.239000   -0.162352  3.693227e+06  5.250000e+05   
25%      1.000000    -0.852500   -0.013927  1.871417e+07  5.000000e+06   
50%      1.000000     0.010000    0.009125  2.740018e+07  7.398704e+06   
75%      1.000000     0.470000    0.031571  4.980786e+07  1.200000e+07   
max      1.000000  3864.500000    0.092896  2.138085e+09  4.212336e+08   

                 C7           T1            T2            T3           T4  \
count    610.000000   681.000000    681.000000    681.000000   681.000000   
mean     500.459962   465.634361  12758.606461  11395.844347   294.353891   
std     1648.337634   175.741647   5449.644597   4839.670179   121.532637   
min        0.074000   132.000000      0.000000      0.000000     0.000000   
25%       37.245750   351.000000   9195.000000   8162.000000   213.000000   
50%      103.833000   444.000000  12045.000000  10785.000000   279.000000   
75%      331.138000   551.000000  15241.000000  13760.000000   354.000000   
max    30683.000000  1750.000000  49056.000000  43952.000000  1058.000000   

                 T5          S1          S2          S3  
count    681.000000  681.000000  681.000000  681.000000  
mean     679.220264   68.421439  120.104258  144.759178  
std      472.914323   39.096525   84.828959   69.276285  
min       -1.000000   -1.000000   20.000000   26.000000  
25%      462.000000   45.000000   73.000000  100.000000  
50%      624.000000   60.000000  100.000000  134.000000  
75%      795.000000   85.000000  142.000000  173.000000  
max    10277.000000  309.000000  944.000000  883.000000

In [5]:
compdata_original.dtypes

I1          object
I2          object
I3          object
P(IPO)     float64
P(H)       float64
P(L)       float64
P(1Day)    float64
C1         float64
C2         float64
C3         float64
C4         float64
C5         float64
C6         float64
C7         float64
T1         float64
T2         float64
T3         float64
T4         float64
T5         float64
S1         float64
S2         float64
S3         float64
dtype: object

In [6]:
compdata_original.median()

P(IPO)     1.350000e+01
P(H)       1.500000e+01
P(L)       1.300000e+01
P(1Day)    1.484500e+01
C1         1.070000e+02
C2         1.000000e+00
C3         1.000000e-02
C4         9.124591e-03
C5         2.740018e+07
C6         7.398704e+06
C7         1.038330e+02
T1         4.440000e+02
T2         1.204500e+04
T3         1.078500e+04
T4         2.790000e+02
T5         6.240000e+02
S1         6.000000e+01
S2         1.000000e+02
S3         1.340000e+02
dtype: float64

In [7]:
# Create a copy of the data
compdata = compdata_original.copy()

# Variable Calculations:

In [8]:
# Impute the rows T2 values!!!! mean
# Cannot do calculations with missing data on the number of words, the ratios will be 0 and create outliers

compdata.loc[compdata['T2'] == 0]


I1                       I2    I3  P(IPO)  P(H)  P(L)  P(1Day)     C1  \
125  CLD  CLOUD PEAK ENERGY INC.   1221    15.0  18.0  16.0    14.84  100.0   
199  EDG        Edgen Group Inc.   5150    11.0  16.0  14.0     9.50  120.0   

      C2    C3  ...          C6        C7     T1   T2       T3     T4      T5  \
125  1.0  2.00  ...  30600000.0  1239.711  656.0  0.0  18173.0  486.0  1135.0   
199  1.0 -0.59  ...  15000000.0   911.612  619.0  0.0  19153.0  474.0  1108.0   

        S1     S2     S3  
125   77.0  229.0  219.0  
199  104.0  224.0  253.0  

[2 rows x 22 columns]

In [9]:
# check for negatives
# cannot have a negative number of long words (T5)

compdata.loc[compdata['T5'] < 0]

I1                          I2    I3  P(IPO)  P(H)  P(L)  P(1Day)    C1  \
7  ACRX  ACELRX PHARMACEUTICALS INC  2834     5.0  14.0  12.0     4.55  91.0   

    C2    C3  ...         C6  C7     T1       T2       T3     T4   T5    S1  \
7  0.0 -0.74  ...  8000000.0 NaN  334.0  11226.0  10056.0  276.0 -1.0  55.0   

      S2     S3  
7  120.0  122.0  

[1 rows x 22 columns]

In [10]:
# cannot have a negative number of positive words (S1)
compdata.loc[compdata['S1'] < 0]

I1            I2    I3  P(IPO)  P(H)  P(L)  P(1Day)     C1   C2    C3  \
134  COGT  Cogent, inc.  7373    12.0  12.0  10.0    17.98  133.0  1.0  0.14   

     ...          C6      C7     T1       T2      T3     T4     T5   S1    S2  \
134  ...  18000000.0  32.179  419.0  10764.0  9856.0  259.0  643.0 -1.0  97.0   

        S3  
134  134.0  

[1 rows x 22 columns]

In [11]:
# Fill the 0, negative, and infinite values with the mean for continuous values
compdata['S1'].mean()

68.42143906020559

In [12]:
#compdata["S1"] = pd.to_numeric(compdata["S1"])

## HELP
s = compdata["S1"]
compdata["S1"].where(s < 0)

## HELP
def no_neg(S1):
    if S1 < 0:
        return(68.42143906020559)
    
compdata['S1'].apply(no_neg)



In [13]:
# compdata.loc[134]

In [14]:
# Ratios:

    # ** How to find the ratio for T1 and T2... impute them

# Percentage of real words
compdata['T3_'] = (compdata['T3']/compdata['T2'])
# Percentage of long sentences
compdata['T4_'] = (compdata['T4']/compdata['T1'])
# Percentage of long words
compdata['T5_'] = (compdata['T5']/compdata['T2'])
# Percentage of positive words
compdata['S1_'] = (compdata['S1']/compdata['T2'])
# Percentage of negative words
compdata['S2_'] = (compdata['S2']/compdata['T2'])
# Percentage of uncertain words
compdata['S3_'] = (compdata['S3']/compdata['T2'])
compdata.head()

# Target & Control Variable Calculation

compdata['P(mid)'] = ((compdata['P(H)']+compdata['P(L)'])/2)

def p(row):
    if row['P(IPO)'] < row['P(mid)']:
        val = 1
    else:
        val = 0
    return val

def q(row):
    if row['P(1Day)'] > row['P(IPO)']:
        val = 1
    else:
        val = 0
    return val

def r(row):
    if row['C3'] >= 0:
        val = 1
    else:
        val = 0
    return val

def s(row):
    if row['P(IPO)'] > row['P(mid)']:
        val = (row['P(IPO)'] - row['P(mid)'])/row['P(mid)']*100
    else:
        val = 0
    return val

compdata['C3_'] = compdata.apply(r, axis=1)
compdata['C5_'] = compdata['C5']/compdata['C6'] 
compdata['C6_'] = compdata.apply(s, axis=1)
compdata['Y1'] = compdata.apply(p, axis=1)
compdata['Y2'] = compdata.apply(q, axis=1)

# Missing Data:

In [15]:
compdata.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 682 entries, 0 to 681
Data columns (total 34 columns):
I1         682 non-null object
I2         682 non-null object
I3         674 non-null object
P(IPO)     677 non-null float64
P(H)       672 non-null float64
P(L)       672 non-null float64
P(1Day)    660 non-null float64
C1         660 non-null float64
C2         660 non-null float64
C3         646 non-null float64
C4         660 non-null float64
C5         676 non-null float64
C6         676 non-null float64
C7         610 non-null float64
T1         681 non-null float64
T2         681 non-null float64
T3         681 non-null float64
T4         681 non-null float64
T5         681 non-null float64
S1         681 non-null float64
S2         681 non-null float64
S3         681 non-null float64
T3_        681 non-null float64
T4_        681 non-null float64
T5_        681 non-null float64
S1_        681 non-null float64
S2_        681 non-null float64
S3_        681 non-null float64
P(

In [16]:
compdata.isna().sum()

I1          0
I2          0
I3          8
P(IPO)      5
P(H)       10
P(L)       10
P(1Day)    22
C1         22
C2         22
C3         36
C4         22
C5          6
C6          6
C7         72
T1          1
T2          1
T3          1
T4          1
T5          1
S1          1
S2          1
S3          1
T3_         1
T4_         1
T5_         1
S1_         1
S2_         1
S3_         1
P(mid)     10
C3_         0
C5_         6
C6_         0
Y1          0
Y2          0
dtype: int64

In [17]:
# what is the percentage of missing data in each column
compdata.isna().mean().round(4) * 100

# Majority of the data is missing from C7

I1          0.00
I2          0.00
I3          1.17
P(IPO)      0.73
P(H)        1.47
P(L)        1.47
P(1Day)     3.23
C1          3.23
C2          3.23
C3          5.28
C4          3.23
C5          0.88
C6          0.88
C7         10.56
T1          0.15
T2          0.15
T3          0.15
T4          0.15
T5          0.15
S1          0.15
S2          0.15
S3          0.15
T3_         0.15
T4_         0.15
T5_         0.15
S1_         0.15
S2_         0.15
S3_         0.15
P(mid)      1.47
C3_         0.00
C5_         0.88
C6_         0.00
Y1          0.00
Y2          0.00
dtype: float64

In [18]:
#isinstance(compdata['I3'], object) 

#compdata['I3'] = [isinstance(x, str) for x in compdata.I3]
#compdata

#compdata[compdata.I3.apply(lambda x: isinstance(x, str))]

In [19]:
compdata[compdata["I1"]=='DLB']

I1                         I2                      I3  P(IPO)  P(H)  \
174  DLB  Dolby Laboratories, Inc.   6794, 3861, 3663, 7819    18.0  18.0   

     P(L)    P(1Day)    C1   C2    C3  ...       T5_       S1_      S2_  \
174  16.0  24.299999  90.0  1.0  0.47  ...  0.059655  0.004513  0.00798   

          S3_  P(mid)  C3_       C5_       C6_  Y1  Y2  
174  0.010793    17.0    1  3.540441  5.882353   0   1  

[1 rows x 34 columns]

In [20]:
compdata[compdata["I1"]=='DTSI']

I1         I2                I3  P(IPO)  P(H)  P(L)  P(1Day)    C1  \
187  DTSI  DTS, INC.  3651, 6794, 7819    17.0  16.0  14.0    24.92  76.0   

      C2    C3  ...       T5_       S1_       S2_      S3_  P(mid)  C3_  \
187  0.0  0.99  ...  0.050882  0.005922  0.011482  0.00846    15.0    1   

          C5_        C6_  Y1  Y2  
187  3.410759  13.333333   0   1  

[1 rows x 34 columns]

In [21]:
compdata[compdata["I1"]=='RLD']

I1          I2                I3  P(IPO)  P(H)  P(L)  P(1Day)    C1  \
499  RLD  RealD Inc.  3663, 3861, 6794    16.0  15.0  13.0    19.51  98.0   

      C2    C3  ...       T5_       S1_      S2_       S3_  P(mid)  C3_  \
499  1.0 -0.33  ...  0.056527  0.006539  0.00874  0.014273    14.0    0   

         C5_        C6_  Y1  Y2  
499  3.81107  14.285714   0   1  

[1 rows x 34 columns]

In [22]:
compdata[compdata["I1"]=='UPG']

I1                      I2          I3  P(IPO)  P(H)  P(L)  P(1Day)  \
620  UPG  UNIVERSAL POWER GROUP   7389, 5063     7.0   9.0   7.0     7.35   

        C1   C2    C3  ...      T5_     S1_       S2_      S3_  P(mid)  C3_  \
620  101.0  0.0  0.38  ...  0.04157  0.0042  0.008835  0.00956     8.0    1   

          C5_  C6_  Y1  Y2  
620  1.666667  0.0   1   1  

[1 rows x 34 columns]

In [23]:
compdata.I3.replace(to_replace ="6794, 3861, 3663, 7819", value ="3000", inplace=True)

In [24]:
compdata.I3.replace(to_replace ="3651, 6794, 7819", value ="9900", inplace=True)

In [25]:
compdata.I3.replace(to_replace ="3663, 3861, 6794", value ="3000", inplace=True)

In [26]:
compdata.I3.replace(to_replace ="7389, 5063", value ="9900", inplace=True)

In [27]:
compdata.I3.replace(to_replace ="541990", value ="7000", inplace=True)

In [28]:
compdata["I3"] = pd.to_numeric(compdata["I3"])

In [29]:
def assign_I3_bins(I3):
    if I3 >= 100 and I3 <= 999:                    
        return(1)
    if I3 >= 1000 and I3 <= 1499:       
        return(2)
    if I3 >= 1500 and I3 <= 1799:       
        return(3)
    if I3 >= 1800 and I3 <= 1999:      
        return(4)
    if I3 >= 2000 and I3 <= 3999:                  
        return(5)
    if I3 >= 4000 and I3 <= 4999:                  
        return(6)
    if I3 >= 5000 and I3 <= 5199:                 
        return(7)
    if I3 >= 5200 and I3 <= 5999:                  
        return(8)
    if I3 >= 6000 and I3 <= 6799:                 
        return(9)
    if I3 >= 7000 and I3 <= 8999:                  
        return(10)
    if I3 >= 9100 and I3 <= 9729:                  
        return(11)   
    if I3 >= 9900 and I3 <= 9999:                  
        return(12)   
    
    # Creating a new column I3_bins in the dataframe

compdata['I3_bins'] = compdata['I3'].apply(assign_I3_bins)  

compdata.head()

I1                                  I2      I3  P(IPO)  P(H)  P(L)  \
0  AATI  ADVANCED ANALOGIC TECHNOLOGIES INC  3674.0    10.0   9.5   8.5   
1  ABPI     ACCENTIA BIOPHARMACEUTICALS INC  2834.0     8.0  10.0   8.0   
2  ACAD          ACADIA PHARMACEUTICALS INC  2834.0     7.0  14.0  12.0   
3  ACHN       ACHILLION PHARMACEUTICALS INC  2834.0    11.5  16.0  14.0   
4  ACLI     AMERICAN COMMERCIAL LINES INC.   4492.0    21.0  21.0  19.0   

     P(1Day)     C1   C2    C3  ...       S1_       S2_       S3_  P(mid)  \
0  11.870000  122.0  1.0  3.43  ...  0.004875  0.009199  0.010929     9.0   
1   7.250000  259.0  0.0 -1.62  ...  0.003258  0.011105  0.010876     9.0   
2   6.700000   90.0  1.0 -1.24  ...  0.011593  0.006271  0.011403    13.0   
3  12.390000  209.0  1.0 -0.91  ...  0.009686  0.007144  0.013319    15.0   
4  56.599998   80.0  1.0  0.07  ...  0.004518  0.010047  0.011261    20.0   

   C3_        C5_        C6_  Y1  Y2  I3_bins  
0    1   3.864345  11.111111   0   1      5.0  
1    0  12.028832   0.000000   1   0      5.0  
2    0   3.369134   0.000000   1   0      5.0  
3    0   3.299697   0.000000   1   1      5.0  
4    1   3.726269   5.000000   0   1      6.0  

[5 rows x 35 columns]

In [30]:
compdata.isna().sum()

I1          0
I2          0
I3          8
P(IPO)      5
P(H)       10
P(L)       10
P(1Day)    22
C1         22
C2         22
C3         36
C4         22
C5          6
C6          6
C7         72
T1          1
T2          1
T3          1
T4          1
T5          1
S1          1
S2          1
S3          1
T3_         1
T4_         1
T5_         1
S1_         1
S2_         1
S3_         1
P(mid)     10
C3_         0
C5_         6
C6_         0
Y1          0
Y2          0
I3_bins     8
dtype: int64

In [31]:
# Since C2 is a binary value, we will fill the missing data with the mode

compdata['Imputed_C2'] = compdata['C2'].fillna(compdata['C2'].mode()[0])
compdata.head()

I1                                  I2      I3  P(IPO)  P(H)  P(L)  \
0  AATI  ADVANCED ANALOGIC TECHNOLOGIES INC  3674.0    10.0   9.5   8.5   
1  ABPI     ACCENTIA BIOPHARMACEUTICALS INC  2834.0     8.0  10.0   8.0   
2  ACAD          ACADIA PHARMACEUTICALS INC  2834.0     7.0  14.0  12.0   
3  ACHN       ACHILLION PHARMACEUTICALS INC  2834.0    11.5  16.0  14.0   
4  ACLI     AMERICAN COMMERCIAL LINES INC.   4492.0    21.0  21.0  19.0   

     P(1Day)     C1   C2    C3  ...       S2_       S3_  P(mid)  C3_  \
0  11.870000  122.0  1.0  3.43  ...  0.009199  0.010929     9.0    1   
1   7.250000  259.0  0.0 -1.62  ...  0.011105  0.010876     9.0    0   
2   6.700000   90.0  1.0 -1.24  ...  0.006271  0.011403    13.0    0   
3  12.390000  209.0  1.0 -0.91  ...  0.007144  0.013319    15.0    0   
4  56.599998   80.0  1.0  0.07  ...  0.010047  0.011261    20.0    1   

         C5_        C6_  Y1  Y2  I3_bins  Imputed_C2  
0   3.864345  11.111111   0   1      5.0         1.0  
1  12.028832   0.000000   1   0      5.0         0.0  
2   3.369134   0.000000   1   0      5.0         1.0  
3   3.299697   0.000000   1   1      5.0         1.0  
4   3.726269   5.000000   0   1      6.0         1.0  

[5 rows x 36 columns]

In [32]:
compdata.isna().sum()

I1             0
I2             0
I3             8
P(IPO)         5
P(H)          10
P(L)          10
P(1Day)       22
C1            22
C2            22
C3            36
C4            22
C5             6
C6             6
C7            72
T1             1
T2             1
T3             1
T4             1
T5             1
S1             1
S2             1
S3             1
T3_            1
T4_            1
T5_            1
S1_            1
S2_            1
S3_            1
P(mid)        10
C3_            0
C5_            6
C6_            0
Y1             0
Y2             0
I3_bins        8
Imputed_C2     0
dtype: int64

In [33]:
# Drop missing P Values from Continues DF

compdata = compdata.dropna(how='all', subset=['P(1Day)'])

In [34]:
# Create DF with only continuous features

# Drop the imputed variable colmuns
# Drop the P values
# Drop Y1 Y2 **make seperate df
    # Drop T1 and T2?

compdata_cont = compdata.drop(['P(IPO)','P(H)','P(L)','P(1Day)','P(mid)','I1','I2','I3','T3','T4','T5','S1','S2','S3','C2','C5','C6','Imputed_C2','I3_bins','Y1','Y2'], axis=1)
compdata_cont.head()

C1    C3        C4       C7     T1       T2       T3_       T4_  \
0  122.0  3.43  0.029074   51.345  470.0  12719.0  0.908876  0.640426   
1  259.0 -1.62 -0.013352   25.936  791.0  21792.0  0.898724  0.644753   
2   90.0 -1.24  0.020715    7.378  201.0   5262.0  0.909350  0.636816   
3  209.0 -0.91  0.020023    8.526  328.0   8259.0  0.917060  0.539634   
4   80.0  0.07 -0.034895  632.298  572.0  14830.0  0.888469  0.587413   

        T5_       S1_       S2_       S3_  C3_        C5_        C6_  
0  0.054250  0.004875  0.009199  0.010929    1   3.864345  11.111111  
1  0.051395  0.003258  0.011105  0.010876    0  12.028832   0.000000  
2  0.061764  0.011593  0.006271  0.011403    0   3.369134   0.000000  
3  0.061630  0.009686  0.007144  0.013319    0   3.299697   0.000000  
4  0.048550  0.004518  0.010047  0.011261    1   3.726269   5.000000

In [35]:
compdata_cont.isna().sum()

C1      0
C3     24
C4      0
C7     60
T1      0
T2      0
T3_     0
T4_     0
T5_     0
S1_     0
S2_     0
S3_     0
C3_     0
C5_     0
C6_     0
dtype: int64

In [36]:
# Create DF with only categorical features
# drop y1 y2 c2 i3
compdata_cat = compdata.drop(['P(IPO)','P(H)','P(L)','P(1Day)','P(mid)','I3','C2','C1','C3','C4','C5','C6','C7','T1','T2','T3','T4','T5','S1','S2','S3','T3_','T4_','T5_','S1_','S2_','S3_','C5_','C6_','Y1','Y2'], axis=1)
compdata_cat.head()

I1                                  I2  C3_  I3_bins  Imputed_C2
0  AATI  ADVANCED ANALOGIC TECHNOLOGIES INC    1      5.0         1.0
1  ABPI     ACCENTIA BIOPHARMACEUTICALS INC    0      5.0         0.0
2  ACAD          ACADIA PHARMACEUTICALS INC    0      5.0         1.0
3  ACHN       ACHILLION PHARMACEUTICALS INC    0      5.0         1.0
4  ACLI     AMERICAN COMMERCIAL LINES INC.     1      6.0         1.0

In [37]:
targetvals = compdata.drop(['P(IPO)','P(H)','P(L)','P(1Day)','P(mid)','I3','C2','C1','C3','C4','C5','C6','C7','T1','T2','T3','T4','T5','S1','S2','S3','T3_','T4_','T5_','S1_','S2_','S3_','C5_','C6_','I1','I2','C3_','I3_bins','Imputed_C2'], axis=1)
targetvals.head()

Y1  Y2
0   0   1
1   1   0
2   1   0
3   1   1
4   0   1

#### Run evaluation code now.. go back and check the results to see if it improved or not

##### after impute, after min max, after normalization

# Pipeline 3: Min Max --> Normalization --> 3 STD

## Min Max:

In [38]:
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()
print(scaler.fit(compdata_cont))

ValueError: Input contains infinity or a value too large for dtype('float64').

In [ ]:
print(scaler.data_max_)

In [ ]:
print(scaler.transform(compdata_cont))

# Skewness Normalization

In [ ]:
# library and data
import seaborn as sns
import matplotlib.pyplot as plt
 
# plot
f, axes = plt.subplots(3, 4, figsize=(15,11), sharex=False)
sns.distplot( compdata_cont["C1"] , ax=axes[0, 0])
sns.distplot( compdata_cont["C4"] , ax=axes[0, 1])
sns.distplot( compdata_cont["T2"] , ax=axes[0, 2])
sns.distplot( compdata_cont["T3_"] , ax=axes[0, 3])
sns.distplot( compdata_cont["T4_"] , ax=axes[1, 0])
sns.distplot( compdata_cont["T5_"] , ax=axes[1, 1])
sns.distplot( compdata_cont["S1_"] , ax=axes[1, 2])
sns.distplot( compdata_cont["S2_"] , ax=axes[1, 3])
sns.distplot( compdata_cont["S3_"] , ax=axes[2, 0])
sns.distplot( compdata_cont["C3_"] , ax=axes[2, 1])
sns.distplot( compdata_cont["C5_"] , ax=axes[2, 2])
sns.distplot( compdata_cont["C6_"] , ax=axes[2, 3])
plt.show()

In [ ]:
compdata_cont.skew()

# check for infinite numbers in df
# close to 0 is more normal (T4_).. [-.5 - .5]

# log all the values for the continues values *positives
# exponential for the negative continues values

# make it fit better into range after log and exponential (sqrt? log again?)

# is any data good? (between -.5 and .5):
    # C4, T1, T4_, T5_, S3_, C3_
    
# Positive Skew: --> log
    # C1, C7, T2, S1_, S2_, C5_, C6_
    
# Negative Skew: --> exponential
    # T3_
    
    # *** C7 needs to be fixed earlier!!

In [ ]:
# Negative Skew:
compdata_cont['T3_'] = compdata_cont['T3_'].apply(np.exp)
compdata_cont['T3_'].skew()

In [ ]:
compdata_cont['T3_'] = compdata_cont['T3_'].apply(np.exp)
compdata_cont['T3_'].skew()

In [ ]:
# Positive skew
compdata_cont['C1'] = compdata_cont['C1'].apply(np.sqrt)
compdata_cont['C1'].skew()

In [ ]:
# Positive skew
compdata_cont['T2'] = compdata_cont['T2'].apply(np.sqrt)
compdata_cont['T2'].skew()

In [ ]:
# Positive skew
compdata_cont['S1_'] = compdata_cont['S1_'].apply(np.log)
compdata_cont['S1_'].skew()

In [ ]:
compdata_cont['S1_'] = compdata_cont['S1_'].apply(np.sqrt)
compdata_cont['S1_'].skew()

In [ ]:
# Positive skew
compdata_cont['S2_'] = compdata_cont['S2_'].apply(np.log)
compdata_cont['S2_'].skew()

In [ ]:
compdata_cont['S2_'] = compdata_cont['S2_'].apply(np.sqrt)
compdata_cont['S2_'].skew()

In [ ]:
# Positive skew
compdata_cont['C5_'] = compdata_cont['C5_'].apply(np.sqrt)
compdata_cont['C5_'].skew()

In [ ]:
# Positive skew

# ***** ??

compdata_cont['C6_'] = compdata_cont['C6_'].apply(np.log)
compdata_cont['C6_'].skew()

#  3 STD

In [ ]:
for col in compdata_cont.columns:
    u_bound = compdata_cont[col].mean() + 3* compdata_cont[col].std()
    compdata_cont[col][compdata_cont[col] > u_bound] = u_bound

In [ ]:
for col in compdata_cont.columns:
    u_bound = compdata_cont[col].mean() - 3* compdata_cont[col].std()
    compdata_cont[col][compdata_cont[col] < l_bound] = l_bound

## Correlation:

In [ ]:
# Correlation between Variables

index= ['P(IPO)', 'P(H)', 'P(L)', 'P(1Day)','C1','C3','C4','C5','C6','C7','T1','T2','T3','T4','T5','S1','S2','S3']
cols = ['P(IPO)', 'P(H)', 'P(L)', 'P(1Day)','C1','C3','C4','C5','C6','C7','T1','T2','T3','T4','T5','S1','S2','S3']
df = DataFrame(abs(np.random.randn(18,18)), index=index,  columns=cols)
fig, ax = plt.subplots(figsize=(15,15))
sns.heatmap(df, annot=True, cbar=True, cmap='Blues')
plt.title("Correlation between Variables")
plt.show()

# look for the ones over .5 or under -.5.. these are highly correlated ones
# flagging step

#every pair highly correlated need to be seperated into a different df

# make differe
    # 1 pair = 2
    # 2 pair = 4

In [ ]:
# explain why theyre seperated

## do feature selection

# correlation is alwasy between -1 and 1

# merge all 3 parts together categorical back into .. categorical, contin (all the continues made.. not 1), target

# zscore

>>> from sklearn.preprocessing import StandardScaler
>>> data = [[0, 0], [0, 0], [1, 1], [1, 1]]
>>> scaler = StandardScaler()
>>> print(scaler.fit(data))
StandardScaler()
>>> print(scaler.mean_)
[0.5 0.5]
>>> print(scaler.transform(data))
[[-1. -1.]
 [-1. -1.]
 [ 1.  1.]
 [ 1.  1.]]
>>> print(scaler.transform([[2, 2]]))
[[3. 3.]]

- report all we did to change the data, not the data describing that we used to impute and change
- keep count of missing